src/lib/proof-system/pool-recycling.test.ts

```diff
@@ -0,0 +1,271 @@
+ * to ensure WASM memory panics are prevented.
+ */
+
+import { describe, test, expect, beforeEach, afterEach } from '@jest/globals';
+import { PoolHealthCoordinator, WorkerHealthReport } from './pool-health-coordinator';
+
+describe('Pool Recycling System', () => {
+  let coordinator: PoolHealthCoordinator;
+
+  beforeEach(() => {
+    // Clean up any global state
+    delete (globalThis as any).__o1js_pool_recycling;
+    delete (globalThis as any).__o1js_force_pool_shutdown;
+    delete (globalThis as any).poolHealthCoordinator;
+    
+    // Create fresh coordinator for each test
+    coordinator = new PoolHealthCoordinator({
+      criticalMemoryMB: 1000,
+      maxPoolAgeMs: 10000, // 10 seconds for testing
+      gracefulShutdownTimeoutMs: 1000, // 1 second for testing
+      healthCheckIntervalMs: 100 // Fast checks for testing
+    });
+  });
+
+  afterEach(async () => {
+    // Wait for any ongoing recycling to complete
+    if (coordinator) {
+      await coordinator.waitForRecyclingToComplete();
+    }
+    
+    // Clean up global state
+    delete (globalThis as any).__o1js_pool_recycling;
+    delete (globalThis as any).__o1js_force_pool_shutdown;
+    delete (globalThis as any).poolHealthCoordinator;
+  });
+
+  test('should trigger recycling when any worker reports critical memory', async () => {
+    // Simulate worker reporting critical memory
+    const criticalReport: WorkerHealthReport = {
+      workerId: 'worker-1',
+      memoryUsageMB: 1200, // Above critical threshold of 1000MB
+      isMemoryCritical: true,
+      timestamp: Date.now()
+    };
+
+    coordinator.receiveHealthReport(criticalReport);
+
+    // Should trigger recycling immediately
+    expect(coordinator.isPoolRecycling()).toBe(true);
+    
+    // Wait for recycling to complete
+    await coordinator.waitForRecyclingToComplete();
+    expect(coordinator.isPoolRecycling()).toBe(false);
+  });
+
+  test('should not trigger recycling for healthy workers', async () => {
+    // Simulate healthy worker
+    const healthyReport: WorkerHealthReport = {
+      workerId: 'worker-1', 
+      memoryUsageMB: 500, // Below threshold
+      isMemoryCritical: false,
+      timestamp: Date.now()
+    };
+
+    coordinator.receiveHealthReport(healthyReport);
+
+    expect(coordinator.isPoolRecycling()).toBe(false);
+  });
+
+  test('should set global recycling flags during recycling', async () => {
+    // Trigger recycling
+    coordinator.receiveHealthReport({
+      workerId: 'worker-1',
+      memoryUsageMB: 1500,
+      isMemoryCritical: true,
+      timestamp: Date.now()
+    });
+
+    // Should set blocking flag immediately
+    expect((globalThis as any).__o1js_pool_recycling).toBe(true);
+
+    await coordinator.waitForRecyclingToComplete();
+
+    // Flags should be cleared after recycling
+    expect((globalThis as any).__o1js_pool_recycling).toBeUndefined();
+    expect((globalThis as any).__o1js_force_pool_shutdown).toBeUndefined();
```

Comment:
Just as a note: it is okay we use global state here (and everywhere) since we don't support running multiple `o1js`es in parallel.